vms/platformvm/state.go

```diff
@@ -440,30 +439,95 @@ func (vm *VM) GetUTXOs(
 	limit int,
 	paginate bool,
 ) ([]*avax.UTXO, ids.ShortID, ids.ID, error) {
-	if limit <= 0 || limit > maxUTXOsToFetch { // Don't fetch more than [maxUTXOsToFetch]
+	if limit <= 0 || limit > maxUTXOsToFetch {
 		limit = maxUTXOsToFetch
 	}
 
+	if paginate {
+		return vm.getPaginatedUTXOs(db, addrs, startAddr, startUTXOID, limit)
+	}
+	return vm.getAllUTXOs(db, addrs, startAddr, startUTXOID, limit)
+}
+
+func (vm *VM) getPaginatedUTXOs(
+	db database.Database,
+	addrs ids.ShortSet,
+	startAddr state.Marshaller,
+	startUTXOID ids.ID,
+	limit int,
+) ([]*avax.UTXO, ids.ShortID, ids.ID, error) {
 	seen := ids.Set{} // IDs of UTXOs already in the list
 	utxos := make([]*avax.UTXO, 0, limit)
 	lastAddr := ids.ShortEmpty
 	lastIndex := ids.Empty
 	addrsList := addrs.List()
 	ids.SortShortIDs(addrsList)
+
 	for _, addr := range addrsList {
 		start := ids.Empty
 		if comp := bytes.Compare(addr.Bytes(), startAddr.Bytes()); comp == -1 { // Skip addresses before [startAddr]
 			continue
 		} else if comp == 0 {
 			start = startUTXOID
 		}
+
+		utxoIDs, err := vm.getReferencingUTXOs(db, addr.Bytes(), start, limit) // Get IDs of UTXOs to fetch
+		if err != nil {
+			return nil, ids.ShortID{}, ids.ID{}, fmt.Errorf("couldn't get UTXOs for address %s: %w", addr, err)
+		}
+		for _, utxoIDKey := range utxoIDs { // Get the UTXOs
+			if seen.Contains(utxoIDKey) { // already have this UTXO in the list
+				continue
+			}
+			utxo, err := vm.getUTXO(db, utxoIDKey)
+			if err != nil {
+				return nil, ids.ShortID{}, ids.ID{}, fmt.Errorf("couldn't get UTXO %s: %w", utxoIDKey, err)
+			}
+			utxos = append(utxos, utxo)
+			seen.Add(utxoIDKey)
+			lastAddr = addr
+			lastIndex = utxoIDKey
+			limit--
+			if limit <= 0 {
+				break // Found [limit] utxos; stop.
+			}
+		}
+
+	}
+	return utxos, lastAddr, lastIndex, nil
+}
+
+func (vm *VM) getAllUTXOs(
```

Comment:
Do we need the startAddr, startUTXOID, and limit for this call since it's getting all of the UTXOs?

Reply:
We defo don't need the limit as it's always `maxUTXOsToFetch` - And in where it's called internally, `getAllUTXOs` it's starting from empty startAddr, startUTXOID.

I'm happy removing them, but do we want to ? We'll be losing the ability to fetch all utxos given a start addr, utxoid.

If this is not something we'll need in the future, then no problem :)